modules/api/src/main/java/net/countercraft/movecraft/TrackedLocation.java

```diff
@@ -1,57 +1,69 @@
-        this.origin = origin;
-        posX = location.getX() - origin.getX();
-        posY = location.getY() - origin.getY();
-        posZ = location.getZ() - origin.getZ();
+    public TrackedLocation(@NotNull Craft craft, @NotNull MovecraftLocation location) {
+        this.craft = craft;
+        MovecraftLocation midPoint = craft.getHitBox().getMidPoint();
+        offSet = new MovecraftLocation(
+                location.getX() - midPoint.getX(),
+                location.getY() - midPoint.getY(),
+                location.getZ() - midPoint.getZ()
+        );
     }
 
     /**
-     * Moves the origin point of the tracked location
-     */
-    public void translate(int dx, int dy, int dz) {
-        origin = new MovecraftLocation(origin.getX() + dx, origin.getY() + dy, origin.getZ() + dz);
-    }
-
-    /**
-     * Rotates the stored location about the origin point location.
+     * Rotates the stored location about any point.
      * @param rotation A clockwise or counter-clockwise direction to rotate.
+     * @param pivot An absolute position at which to rotate the trackedLocation.
      */
-    public void rotate(MovecraftRotation rotation) {
-        int newX;
-        int newZ;
-        if (rotation == MovecraftRotation.CLOCKWISE) {
-            newX = posZ + origin.getX();
-            newZ = -posX + origin.getZ();
-        } else if (rotation == MovecraftRotation.ANTICLOCKWISE) {
-            newX = -posZ + origin.getX();
-            newZ = posX + origin.getZ();
-        } else return;
-
-        posX = newX;
-        posZ = newZ;
+    public void rotate(MovecraftRotation rotation, MovecraftLocation pivot) {
+        //This is the position vector relative to the tracked location to the pivot.
+        MovecraftLocation offSetToPivot = new MovecraftLocation(
+                getAbsoluteLocation().getX() - pivot.getX(),
+                getAbsoluteLocation().getY() - pivot.getY(),
+                getAbsoluteLocation().getZ() - pivot.getZ()
+        );
+        //This is the position vector relative to the midpoint to the pivot.
+        MovecraftLocation midPoint = craft.getHitBox().getMidPoint();
+        MovecraftLocation midPointToPivot = new MovecraftLocation(
+                midPoint.getX() - pivot.getX(),
+                midPoint.getY() - pivot.getY(),
+                midPoint.getZ() - pivot.getZ()
+        );
+        //Rotates both vectors
+        MovecraftLocation rotatedOffSetToPivot = MathUtils.rotateVec(rotation, offSetToPivot);
+        MovecraftLocation rotatedMidPointToPivot = MathUtils.rotateVec(rotation, midPointToPivot);
+        //Subtracts rotatedOffSetToPivot - rotatedMidPointToPivot to get the new offset to the midpoint.
+        offSet = new MovecraftLocation(
+                rotatedOffSetToPivot.getX() - rotatedMidPointToPivot.getX(),
+                rotatedOffSetToPivot.getY() - rotatedMidPointToPivot.getY(),
+                rotatedOffSetToPivot.getZ() - rotatedMidPointToPivot.getZ()
+        );
```

Comment:
Check how rotation does this, there's a utility function to do all this:
https://github.com/APDevTeam/Movecraft/blob/main/modules/Movecraft/src/main/java/net/countercraft/movecraft/async/rotation/RotationTask.java#L117

Reply:
I wonder, if a turret were to rotate in the craft, would the craft be able to track that as well? Or is that not worth going for right now?